### PR TITLE
956 store paas secrets in azure key vault

### PR DESCRIPTION
**Pull request by felixtheflex**
### Context

Use secrets stored on AzureKey Vault

### Changes proposed in this pull request

Use secrets stored on AzureKey Vault

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally
- [ ] Product Review

### CODE REVIEW

File: terraform/data.tf [hunk 1]
```diff
@@ -0,0 +1,14 @@
+data "azurerm_key_vault" "vault" {
```
Comment (vigneshmsft) [inline terraform/data.tf:1]: quotes can be removed.

File: terraform/terraform.tf [hunk 2]
```diff
@@ -20,16 +20,28 @@ terraform {
...
 
 provider statuscake {
-  username = var.statuscake_username
-  apikey   = var.statuscake_password
+  username = local.infra_secrets.STATUSCAKE_USERNAME
+  apikey   = local.infra_secrets.STATUSCAKE_PASSWORD
 }
 
+provider azurerm {
+    version = "2.45.1"
```
Comment (vigneshmsft) [inline terraform/terraform.tf:34]: its not required here, as its specified in the top section

File: terraform/terraform.tf [hunk 3]
```diff
@@ -48,3 +60,13 @@ module statuscake {
 
   alerts = var.statuscake_alerts
 }
+
+#module keyvault {
```
Comment (vigneshmsft) [inline terraform/terraform.tf]: pls remove commented code.

File: terraform/variables.tf [hunk 4]
```diff
@@ -39,11 +47,16 @@ variable statuscake_password { default = "not-empty" }
```
Comment (vigneshmsft) [inline terraform/variables.tf:47]: pls remove statuscake_username, statuscake_password, dockerhub_username, dockerhub_password, cf_user, cf_password variables

File: terraform/variables.tf [hunk 5]
```diff
@@ -39,11 +47,16 @@ variable statuscake_password { default = "not-empty" }
 
 locals {
   cf_api_url = "https://api.london.cloud.service.gov.uk"
+
+
+  paas_app_config                = yamldecode(file(var.paas_app_config_file))[var.paas_app_environment_config]
+  app_secrets                    = yamldecode(data.azurerm_key_vault_secret.app_secrets.value)
+  infra_secrets                  = yamldecode(data.azurerm_key_vault_secret.infra_secrets.value)
+  paas_app_environment_variables = merge(local.app_secrets, local.paas_app_config)
+  azure_credentials = jsondecode(var.azure_credentials)
```
Comment (vigneshmsft) [inline terraform/variables.tf:56]: `terraform fmt -recursive` and remove extra blank line #50

**Comment by felixtheflex** [comment]
[Publish: Move secrets to keyvault](https://trello.com/c/VFFJrFuF/956-publish-move-secrets-to-keyvault)

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -86,13 +86,9 @@ jobs:
...
         env:
           ARM_ACCESS_KEY:             ${{ secrets[format('ARM_ACCESS_KEY_{0}', env.DEPLOY_ENV)] }}
-          TF_VAR_dockerhub_username:  ${{ secrets.DOCKERHUB_USERNAME }}
-          TF_VAR_dockerhub_password:  ${{ secrets.DOCKERHUB_PASSWORD }}
           TF_VAR_paas_docker_image:   ${{ env.DOCKER_IMAGE }}
-          TF_VAR_cf_user:             ${{ secrets[format('CF_USERNAME_{0}', env.DEPLOY_ENV)] }}
-          TF_VAR_cf_user_password:    ${{ secrets[format('CF_PASSWORD_{0}', env.DEPLOY_ENV)] }}
-          TF_VAR_statuscake_username: ${{ secrets.STATUSCAKE_USERNAME }}
-          TF_VAR_statuscake_password: ${{ secrets.STATUSCAKE_API_KEY }}
+          TF_VAR_azure_credentials:   ${{ secrets[format('AZURE_CREDENTIALS_{0}', matrix.environment)] }}
```
Comment (vigneshmsft) [inline .github/workflows/deploy.yml:90]:
```suggestion
          TF_VAR_azure_credentials:   ${{ secrets[format('AZURE_CREDENTIALS_{0}', env.DEPLOY_ENV)] }}
```
`env.DEPLOY_ENV` to keep it consistent with the other variables.

File: terraform/workspace_variables/production.tfvars [hunk 7]
```diff
@@ -18,3 +18,8 @@ statuscake_alerts = {
     node_locations = ["UKINT", "UK1", "MAN1", "MAN5", "DUB2"]
   }
 }
+
+key_vault_name           = "s121d01-shared-kv-01"
+key_vault_resource_group = "s121d01-shared-rg"
+key_app_secret_name      = "PUBLISH-APP-SECRETS-PRODUCTION"
+key_infra_secret_name    = "BAT-INFRA-SECRETS-PRODUCTION"
```
Comment (vigneshmsft) [inline terraform/workspace_variables/production.tfvars:25]: key_vault_name and resource_group values are incorrect for the env, it should be `p01` 
same in other files as well

also can you rename `key_app_secret_name` and `key_infra_secret_name` to
`key_vault_app_secret_name ` and `key_vault_infra_secret_name ` to keep it consistent?

File: terraform/data.tf [hunk 8]
```diff
@@ -0,0 +1,14 @@
+data azurerm_key_vault vault {
+  name                = var.key_vault_name           #keyvault name
```
Comment (saliceti) [inline terraform/data.tf:2]: The comments are not particularly useful in this file

File: terraform/workspace_variables/staging.tfvars [hunk 9]
```diff
@@ -18,3 +18,9 @@ statuscake_alerts = {
     node_locations = ["UKINT", "UK1", "MAN1", "MAN5", "DUB2"]
   }
 }
+
+#vault
+key_vault_name           = "s121d01-shared-kv-01"
```
Comment (saliceti) [inline terraform/workspace_variables/staging.tfvars:23]: Should be s121t01

File: terraform/workspace_variables/sandbox.tfvars [hunk 10]
```diff
@@ -18,3 +18,8 @@ statuscake_alerts = {
     node_locations = ["UKINT", "UK1", "MAN1", "MAN5", "DUB2"]
   }
 }
+
+key_vault_name           = "s121d01-shared-kv-01"
```
Comment (saliceti) [inline terraform/workspace_variables/sandbox.tfvars:22]: I would have assumed same as prod s121p01 ?

File: .gitignore [hunk 11]
```diff
@@ -53,5 +53,6 @@ terraform.tfstate.d/
 lock.json
 **/.terraform/
 terraform/**/*.yml
+terraform/**/*.json
```
Comment (saliceti) [inline .gitignore:56]: These shouldn't be required anymore:
```
terraform/**/*.yml
terraform/**/*.json
terraform/**/*.base64
!terraform/workspace_variables/app_config.yml
```

File: terraform/workspace_variables/sandbox.tfvars (hunk 10 above)
Comment (felixtheflex) [reply terraform/workspace_variables/sandbox.tfvars]: correct... done.

File: terraform/workspace_variables/staging.tfvars (hunk 9 above)
Comment (felixtheflex) [reply terraform/workspace_variables/staging.tfvars]: done..

File: terraform/data.tf (hunk 8 above)
Comment (felixtheflex) [reply terraform/data.tf]: done...

File: terraform/workspace_variables/production.tfvars (hunk 7 above)
Comment (felixtheflex) [reply terraform/workspace_variables/production.tfvars]: ok... done.

File: .gitignore (hunk 11 above)
Comment (felixtheflex) [reply .gitignore:56]: I agree.... but could leave until after this PR.

**Review by vigneshmsft — APPROVED** [review]

File: terraform/workspace_variables/sandbox.tfvars [hunk 12]
```diff
@@ -18,3 +18,8 @@ statuscake_alerts = {
     node_locations = ["UKINT", "UK1", "MAN1", "MAN5", "DUB2"]
   }
 }
+
+key_vault_name           = "s121p01-shared-kv-01"
+key_vault_resource_group = "s121p01-shared-rg"
+key_vault_app_secret_name      = "PUBLISH-APP-SECRETS-PRODUCTION"
```
Comment (saliceti) [inline terraform/workspace_variables/sandbox.tfvars:24]: The sandbox secrets may be different than the prod ones. Shall we create a separate sandbox one?

File: terraform/workspace_variables/staging.tfvars [hunk 13]
```diff
@@ -18,3 +18,9 @@ statuscake_alerts = {
     node_locations = ["UKINT", "UK1", "MAN1", "MAN5", "DUB2"]
   }
 }
+
+#vault
+key_vault_name           = "s121t01-shared-kv-01"
```
Comment (saliceti) [inline terraform/workspace_variables/staging.tfvars:23]: The alignment doesn't look correct. Have you run terraform fmt?

Comment (felixtheflex) [reply terraform/workspace_variables/staging.tfvars:23]: all ready done..

Comment (saliceti) [reply terraform/workspace_variables/staging.tfvars]: Please run it again then

File: terraform/workspace_variables/sandbox.tfvars (hunk 12 above)
Comment (vigneshmsft) [reply terraform/workspace_variables/sandbox.tfvars]: yes, have already created `PUBLISH-APP-SECRETS-SANDBOX` & `BAT-INFRA-SECRETS-SANDBOX`

File: .gitignore (hunk 11 above)
Comment (felixtheflex) [reply .gitignore]: all done....

**Review by saliceti — APPROVED** [review]

**Pull request merged**